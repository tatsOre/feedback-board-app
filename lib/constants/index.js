export const ROADMAP_SECTION_TAGLINES = {
  planned: 'Ideas priorized for research',
  'in-progress': 'Features currently being developed',
  live: 'Released features',
}

export const FILTER_OPTIONS = [
  { label: 'All', value: 'all' },
  { label: 'UI', value: 'ui' },
  { label: 'UX', value: 'ux' },
  { label: 'Enhancement', value: 'enhancement' },
  { label: 'Bug', value: 'bug' },
  { label: 'Feature', value: 'feature' },
]

export const SORT_OPTIONS = [
  { label: 'Most Upvotes', value: 'MOST_UPVOTES' },
  { label: 'Least Upvotes', value: 'LEAST_UPVOTES' },
  { label: 'Most Comments', value: 'MOST_COMMENTS' },
  { label: 'Least Comments', value: 'LEAST_COMMENTS' },
]

export const COLOR_THEME = {
  primary: {
    base: 'bg-violet-900',
    hover: 'hover:bg-violet-500',
    disabled: 'bg-violet-500',
  },
  secondary: { base: 'bg-indigo-800', hover: 'hover:bg-indigo-500' },
  tertiary: { base: 'bg-blue-900', hover: 'hover:bg-blue-500' },
  info: { base: 'bg-blue-500', hover: 'hover:bg-blue-900' },
  danger: { base: 'bg-red-900', hover: 'hover:bg-red-500' },
}

export const CATEGORY_OPTIONS = [
  { label: 'Feature', value: 'feature' },
  { label: 'UI', value: 'ui' },
  { label: 'UX', value: 'ux' },
  { label: 'Enhancement', value: 'enhancement' },
  { label: 'Bug', value: 'bug' },
]

export const STATUS_OPTIONS = [
  { label: 'Suggestion', value: 'suggestion' },
  { label: 'Planned', value: 'planned' },
  { label: 'In-Progress', value: 'in-progress' },
  { label: 'Live', value: 'live' },
]
