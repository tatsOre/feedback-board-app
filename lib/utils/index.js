export const toCapitalize = (text) => {
  if (!text || typeof text !== 'string') return ''
  let label = text
  if (text.includes('-')) label = text.replace('-', ' ')
  return label
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ')
}

export const getCommentsLength = (comments) => {
  if (!comments || !comments.length) return 0
  return comments.reduce((acc, comment) => {
    return acc + (comment.replies?.length || 0)
  }, comments.length || 0)
}

export const getRandomNumber = (min, max) => {
  let difference = max - min

  let rand = Math.random()

  rand = Math.floor(rand * difference)

  return rand + min
}

export const reduceFeedbacksData = (data) => {
  const accSchema = {
    planned: [],
    'in-progress': [],
    live: [],
    suggestion: [],
  }

  const result = data.reduce((acc, req) => {
    let status = req.status
    let doc = {
      id: req._id.toString(),
      title: req.title,
      slug: req.slug,
      description: req.description,
      status: req.status,
      category: req.category,
      comments: getCommentsLength(req.comments),
      upvotes: req.upvotes,
    }
    acc[status] = acc[status].length ? [...acc[status], doc] : [doc]
    return acc
  }, accSchema)

  return result
}
