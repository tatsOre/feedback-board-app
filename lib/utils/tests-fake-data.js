import MOCK_FEEDBACKS_DATA from '../../__mocks__/mock-feedbacks.json'
import MOCK_USER_DATA from '../../__mocks__/mock-user-info.json'
import MOCK_USER_FEEDBACK_DATA from '../../__mocks__/mock-user-feedback.json'
import MOCK_DETAILED_FEEDBACK_DATA from '../../__mocks__/mock-detailed-feedbacks.json'


export const MOCK_USER = MOCK_USER_DATA.data

export const MOCK_USER_FEEDBACK = MOCK_USER_FEEDBACK_DATA.data

export const MOCK_FEEDBACKS = MOCK_FEEDBACKS_DATA.data

export const MOCK_PLANNED = MOCK_FEEDBACKS.planned

export const MOCK_IN_PROGRESS = MOCK_FEEDBACKS['in-progress']

export const MOCK_LIVE = MOCK_FEEDBACKS.live

export const MOCK_SUGGESTION = MOCK_FEEDBACKS.suggestion

export const MOCK_DETAILED_FEEDBACK = MOCK_DETAILED_FEEDBACK_DATA.data


// import '@testing-library/jest-dom'