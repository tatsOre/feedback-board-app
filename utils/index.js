export const toCapitalize = (text) => {
  if (!text || typeof text !== 'string') return ''
  let label = text
  if (text.includes('-')) label = text.replace('-', ' ')
  return label
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ')
}

export const getCommentsLength = (comments) => {
  if (!comments.length) return 0
  return comments.reduce((acc, comment) => {
    return acc + (comment.replies?.length || 0)
  }, comments.length || 0)
}

export const getRandomNumber = (min, max) => {
  let difference = max - min

  let rand = Math.random()

  rand = Math.floor(rand * difference)

  return rand + min
}
