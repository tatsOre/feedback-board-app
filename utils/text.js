export const toCapitalize = (text) => {
  if (!text || typeof text !== "string") return "";

  let label = text;
  if (text.includes("-")) label = text.replace("-", " ");

  return label
    .split(" ")
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(" ");
};
