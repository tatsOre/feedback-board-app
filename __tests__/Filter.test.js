import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import FilterTags from '../components/Filter'
import { FILTER_OPTIONS as ITEMS } from '../lib/constants'
import { getRandomNumber } from '../lib/utils'

describe('<FilterTags />', () => {
  const initialState = ITEMS[0].value

  const handleChange = jest.fn()

  it('renders with the selected option passed via props', () => {
    const { container } = render(
      <FilterTags
        items={ITEMS}
        checked={initialState}
        onChange={handleChange}
      />
    )

    expect(container).toMatchSnapshot()

    const checked = screen.getByRole('radio', { checked: true })

    expect(checked.value).toEqual(initialState)

    expect(screen.getAllByRole('radio', { checked: false })).toHaveLength(
      ITEMS.length - 1
    )
  })

  it('calls onChange prop when clicked', async () => {
    render(
      <FilterTags
        items={ITEMS}
        checked={initialState}
        onChange={handleChange}
      />
    )
    const index = getRandomNumber(1, ITEMS.length)

    const input = screen.getByLabelText(ITEMS[index].label)

    await userEvent.click(input)

    expect(handleChange).toHaveBeenCalledTimes(1)
  })

  it('rerenders the updated props correctly', () => {
    const { rerender } = render(
      <FilterTags
        items={ITEMS}
        checked={initialState}
        onChange={handleChange}
      />
    )
    const index = getRandomNumber(1, ITEMS.length)

    const selected = screen.getByLabelText(ITEMS[index].label, {
      selector: 'input',
    })

    rerender(
      <FilterTags
        items={ITEMS}
        checked={selected.value}
        onChange={handleChange}
      />
    )

    const checked = screen.getByRole('radio', { checked: true })

    expect(checked.value).toEqual(ITEMS[index].value)

    expect(screen.getAllByRole('radio', { checked: false })).toHaveLength(
      ITEMS.length - 1
    )
  })
})
