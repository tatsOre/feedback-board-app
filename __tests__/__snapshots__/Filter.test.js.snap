// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterTags /> renders with the selected option passed via props 1`] = `
<div>
  <fieldset
    class="bg-white rounded-10 w-full md:w-1/2 lg:w-full p-5"
  >
    <legend
      class="invisible h-0"
    >
      Filter by category:
    </legend>
    <input
      checked=""
      id="filter-option-all"
      name="category"
      tabindex="1"
      type="radio"
      value="all"
    />
    <label
      class="text-white bg-blue-900 rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-all"
    >
      All
    </label>
    <input
      id="filter-option-ui"
      name="category"
      tabindex="2"
      type="radio"
      value="ui"
    />
    <label
      class="text-blue-900 bg-indigo-100 hover:bg-indigo-300 cursor-pointer rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-ui"
    >
      UI
    </label>
    <input
      id="filter-option-ux"
      name="category"
      tabindex="3"
      type="radio"
      value="ux"
    />
    <label
      class="text-blue-900 bg-indigo-100 hover:bg-indigo-300 cursor-pointer rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-ux"
    >
      UX
    </label>
    <input
      id="filter-option-enhancement"
      name="category"
      tabindex="4"
      type="radio"
      value="enhancement"
    />
    <label
      class="text-blue-900 bg-indigo-100 hover:bg-indigo-300 cursor-pointer rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-enhancement"
    >
      Enhancement
    </label>
    <input
      id="filter-option-bug"
      name="category"
      tabindex="5"
      type="radio"
      value="bug"
    />
    <label
      class="text-blue-900 bg-indigo-100 hover:bg-indigo-300 cursor-pointer rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-bug"
    >
      Bug
    </label>
    <input
      id="filter-option-feature"
      name="category"
      tabindex="6"
      type="radio"
      value="feature"
    />
    <label
      class="text-blue-900 bg-indigo-100 hover:bg-indigo-300 cursor-pointer rounded-10 inline-block px-4 py-2 md:py-1.5 mr-2 mb-3 text-sm font-semibold"
      for="filter-option-feature"
    >
      Feature
    </label>
  </fieldset>
</div>
`;
